pkg/csr/csr.go

```diff
@@ -0,0 +1,219 @@
+			break
+		}
+	}
+	if !hasMatch {
+		return false, errors.Errorf("CSR is not from an authorized source %s", csr.Name)
+	}
+
+	if isNodeClientCert(csr, parsedCSR) {
+		// this is done before the instance becomes a node
+		return c.authorizeNodeClientCSR(dnsAddr)
+	}
+
+	return c.authorizeKubeletServingCSR(csr, parsedCSR)
+}
+
+// authorizeKubeletServingCSR authorizes a kubelet serving CSR
+func (c *Config) authorizeKubeletServingCSR(req *certificatesv1.CertificateSigningRequest, csr *x509.CertificateRequest) (bool, error) {
+	if req == nil || csr == nil {
+		return false, errors.New("CSR should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(req.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR contains too few groups: %s", req.Name)
+	}
+	groupSet := sets.NewString(req.Spec.Groups...)
+	if !groupSet.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR does not contain required groups: %s", req.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !hasExactUsages(req, kubeletServerUsages) {
+		return false, errors.Errorf("CSR does not contain required usages: %s", req.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR does not contain required organization: %s", req.Name)
+	}
+	return true, nil // Approve node server cert
+}
+
+// authorizeNodeClientCSR validates a kubelet client CSR by verifying if it already
+// exists as a node because client CSR's are issued by an instance before becoming a node.
+func (c *Config) authorizeNodeClientCSR(nodeName string) (bool, error) {
+	if nodeName == "" {
+		return false, errors.New("node name should not be empty")
+	}
+	_, err := c.k8sClientSet.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, errors.Wrapf(err, "unable to get node %s", nodeName)
+	} else if err == nil {
+		return false, errors.Wrapf(err, "%s node already exists, cannot Approve", nodeName)
+	}
+	return true, nil // Approve node client cert
+}
+
+// Approve approves a CSR by updating it's status conditions
+func (c *Config) Approve(csr *certificatesv1.CertificateSigningRequest) error {
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:           certificatesv1.CertificateApproved,
+		Status:         "True",
+		Message:        "This CSR was approved by the WMCO CSR Approver",
+		LastUpdateTime: metav1.Now(),
+		Reason:         "WMCOApproved",
+	})
+	if _, err := c.k8sClientSet.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		csr.Name, csr, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// isNodeClientCert returns true if the CSR is a kubelet client CSR
```

Comment:
Please link to documentation where this is defined i.e. how did you find out what makes a valid kubelet client CSR?